### PR TITLE
fix: Codegen issues re: recursion, Swift keywords in unions

### PR DESCRIPTION
**Pull request by jbelkins**
## Issue \#
https://github.com/awslabs/aws-sdk-swift/issues/1234 (Recursive unions)
https://github.com/awslabs/aws-sdk-swift/issues/1235 (Reserved word)
https://github.com/awslabs/aws-sdk-swift/issues/1096 (Recursive unions protocol test)

## Description of changes
Fixes two codegen issues that prevented models from generating valid Swift.
- Recursive unions were not being marked with the `indirect` keyword when Swift requires it.
- Swift reserved words appearing in the cases of Smithy unions (i.e. Swift enums with associated types) resulted in invalid Swift

Also fixes an issue which arose when the recursive unions protocol test was re-enabled: that protocol test suite includes a model with sparse timestamps, which were not generating valid Swift in their deserializer method.  That code is now fixed, though it is not used in protocol tests.

## Scope
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SymbolVisitor.kt [hunk 1]
```diff
@@ -97,7 +97,7 @@ class SymbolVisitor(private val model: Model, swiftSettings: SwiftSettings) :
     override fun toMemberName(shape: MemberShape): String {
         val containingShape = model.expectShape(shape.container)
         if (containingShape is UnionShape) {
-            val name = escaper.escapeMemberName(shape.memberName)
+            val name = escaper.escapeMemberName(shape.memberName.toLowerCamelCase())
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SymbolVisitor.kt:100]: This was preventing the union member named `Enum` from being properly escaped in Swift code when it is downcased to the Swift keyword `enum`.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/customtraits/RecursiveUnionTrait.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package software.amazon.smithy.swift.codegen.customtraits
+
+import software.amazon.smithy.model.node.Node
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.model.traits.Trait
+
+class RecursiveUnionTrait : Trait {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/customtraits/RecursiveUnionTrait.kt:7]: This trait is added during preprocessing to a union that should be marked `indirect`.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/serde/json/MemberShapeDecodeGenerator.kt [hunk 3]
```diff
@@ -179,11 +179,21 @@ abstract class MemberShapeDecodeGenerator(
                     } else { // decode date as a string manually
                         val dateName = "date$level"
                         val swiftTimestampName = TimestampHelpers.generateTimestampFormatEnumValue(timestampFormat)
-                        writer.write(
-                            "let \$L = try containerValues.timestampStringAsDate(\$L, format: .\$L, forKey: .\$L)",
-                            dateName, iteratorName, swiftTimestampName, topLevelMember.memberName
-                        )
-                        writer.write("${decodedMemberName}$terminator.$insertMethod($dateName)")
+                        if (!isSparse) {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/serde/json/MemberShapeDecodeGenerator.kt:182]: Timestamps were not being decoded properly when the source string comes from a sparse array.  In that case, the source is now properly unwrapped before conversion to a Swift `Date`.

File: smithy-swift-codegen/src/test/kotlin/UnionGeneratorTests.kt [hunk 4]
```diff
@@ -111,11 +112,13 @@ class UnionGeneratorTests {
         )
         val simpleUnionShape = simpleUnionShapeBuilder.build()
         val model = createModelFromShapes(simpleUnionShape)
-        val settings = model.defaultSettings()
-        val provider: SymbolProvider = SwiftCodegenPlugin.createSymbolProvider(model, settings)
+        val transformedModel = UnionIndirectivizer.transform(model)
```
Comment (jbelkins) [inline smithy-swift-codegen/src/test/kotlin/UnionGeneratorTests.kt:115]: Had to run this transformer on the model so the test would pass.
Another example of how our test setup needs to be unified with our production logic.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/UnionIndirectivizer.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import software.amazon.smithy.model.shapes.MapShape
+import software.amazon.smithy.model.shapes.SetShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.shapes.UnionShape
+import software.amazon.smithy.model.transform.ModelTransformer
+import software.amazon.smithy.swift.codegen.customtraits.RecursiveUnionTrait
+import software.amazon.smithy.swift.codegen.customtraits.SwiftBoxTrait
+
+object UnionIndirectivizer {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/UnionIndirectivizer.kt:16]: Based on `RecursiveShapeBoxer` but just for detecting unions.

**Review by dayaffe — APPROVED** [review]

**Comment by dayaffe** [comment]
@jbelkins What do you think about adding a codegen test that starts with a smithy file?

**Pull request merged**